src/adams_methods.jl

```diff
@@ -0,0 +1,655 @@
+
+        # Shift data from this step, from the "current step" position to the
+        # "previous step" position for ϕstar and ϕ
+        # TODO: implement with DataStructures.CircularBuffer
+        ϕstar[1,1:min(max(n-1,1),current_steporder+1)] =
+                                ϕstar[2,1:min(max(n-1,1),current_steporder+1)]
+        ϕ[1,1:min(max(n-1,1),current_steporder+1)] =
+                                ϕ[2,1:min(max(n-1,1),current_steporder+1)]
+
+        # Update stepsize pointers
+        prev_h = h
+        h = next_h
+        # Increment our step
+        n= n+1
+    end #while loop over steps
+
+    if points == :steps
+        return vcat(t), y
+    else
+        return vcat(tout), yout
+    end
+end
+
+###############################################################################
+#
+# HELPER FUNCTIONS and TABLEAUS FOR ADAMS SOLVERS
+#
+###############################################################################
+##Adams Bashforth Coefficients for Explicit Method
+const ms_coefficients4 = Float64[1        0        0         0
+                                         -1//2    3//2     0         0
+                                         5//12    -4//3    23//12    0
+                                         -9//24   37//24   -59//24   55//24]
+
+##Adams Moulton Coefficients for Implicit Method
+const am_imp_coefficients3 = Float64[1         0        0         0
+                                            1//2      1//2     0         0
+                                            -1//12    8//12    5//12     0
+                                            1//24     -5//24   19//24    9//24]
+
+
+
+###############################################################################
+# Gamma star function from Hairer et al pg 359, table 1.2
+# used in adaptive step selection (See Harier pg 423)
+###############################################################################
+"compute γ*(j) for ode113"
+function γ_star(j)
+    if j == 0
+        return 1
+    elseif j > 0
+
+    return Polynomials.polyval((1 / factorial(j))*
+            Polynomials.polyint(Polynomials.poly(diagm(collect(1-(j-1):1)))),1)
+    else
+        error("order must be positive integer")
+    end
+end
+
+################################################################################
+# Helper functions for calculating coefficients from recurence relationships
+################################################################################
+#Based on Harier Lemma 5.1 on page 399
+#Calculates c_j,q and g_n,j for n and j =1:steporder
+"compute g_{n,j} coefficients for ode113"
+function g_coefs!(g,c,h,t,t_np1,n,steporder)
+    for j = 1:steporder
+      for q = 1:(steporder)-(j-1)
+        if j ==1
+          c[j,q]=1/q
+        elseif j==2
+          c[j,q]=1/q/(q+1)
+        else
+          c[j,q] = c[j-1,q] - c[j-1,q+1]*h/(t_np1-t[n-(j-1)+1])
+        end
+      end
+      g[j]=c[j,1]
+    end
```

Comment:
I like to have an explicit `return nothing` statement in functions like this and the ones below.